pkg/pillar/vault/handler_zfs.go

```diff
@@ -0,0 +1,256 @@
+		h.log.Errorf("Error unlocking vault: %v", err)
+		return err
+	}
+
+	return nil
+}
+
+// e.g. zfs create -o encryption=aes-256-gcm -o keylocation=file://tmp/raw.key -o keyformat=raw persist/vault
+func (h *ZFSHandler) createVault(vaultPath string) error {
+	// prepare key in the staging file
+	// we never create deprecated vault on ZFS
+	// cloudKeyOnlyMode=false, useSealedKey=true
+	unstage, err := stageKey(h.log, false, true, h.options.TpmKeyOnlyMode, zfsKeyDir, zfsKeyFile)
+	if err != nil {
+		return err
+	}
+	defer unstage()
+
+	if err := zfs.CreateVaultDataset(vaultPath, zfsKeyFile); err != nil {
+		h.log.Errorf("Error creating zfs vault %s, error=%v", vaultPath, err)
+		return err
+	}
+
+	h.log.Functionf("Created new vault %s", vaultPath)
+	return nil
+}
+
+// e.g. zfs get keystatus persist/vault
+func (h *ZFSHandler) checkZfsKeyStatus(vaultPath string) error {
+	if _, err := zfs.GetDatasetKeyStatus(vaultPath); err != nil {
+		h.log.Tracef("keystatus query for %s results in error=%v",
+			vaultPath, err)
+		return err
+	}
+
+	return nil
+}
+
+func (h *ZFSHandler) setupVault(vaultPath string) error {
+	// zfs get keystatus returns success as long as vaultPath is a dataset,
+	// (even if not mounted yet), so use it to check dataset presence
+	if err := h.checkZfsKeyStatus(vaultPath); err == nil {
+		//present, call unlock
+		return h.unlockVault(vaultPath)
+	}
+	// If it does not exist then the mount presumbly does not exist either but
+	// double check.
+	if !isDirEmpty(h.log, vaultPath) {
+		h.log.Noticef("Not disturbing non-empty vault(%s)",
+			vaultPath)
+	} else {
+		h.log.Warnf("Clear saved keys for empty vault(%s)",
+			vaultPath)
+		if err := etpm.WipeOutStaleSealedKeyIfAny(); err != nil {
+			h.log.Errorf("WipeOutStaleSealKeyIfAny failed: %s", err)
+		}
+	}
+	// try creating the dataset
+	return h.createVault(vaultPath)
+}
+
+func (h *ZFSHandler) getVaultStatus(vaultName, vaultPath string) *types.VaultStatus {
+	status := types.VaultStatus{}
+	status.Name = vaultName
+	zfsEncryptStatus, zfsEncryptError := h.GetOperationalInfo()
+	if zfsEncryptStatus != info.DataSecAtRestStatus_DATASEC_AT_REST_ENABLED {
+		status.Status = zfsEncryptStatus
+		status.SetErrorDescription(types.ErrorDescription{Error: zfsEncryptError})
+	} else {
+		if err := h.checkOperationalStatus(vaultPath); err != nil {
+			h.log.Errorf("Status failed, %s", err)
+			status.Status = info.DataSecAtRestStatus_DATASEC_AT_REST_ERROR
+			status.SetErrorDescription(types.ErrorDescription{Error: err.Error()})
+		} else {
+			h.log.Functionf("checkOperStatus returns ok for %s", vaultPath)
+			status.Status = info.DataSecAtRestStatus_DATASEC_AT_REST_ENABLED
+		}
+	}
+	return &status
+}
+
+// checkOperationalStatus returns nil if for vaultPath properties
+// for in ZFS have the following values:
+// mounted:yes keystatus:available encryption:aes-256-gcm;
+// else return err
+func (h *ZFSHandler) checkOperationalStatus(vaultPath string) error {
```

Comment:
Not so long ago I already updated this function, but did not dare to move it from vaultzfs.go to the zfs package.  I think it makes sense now.